### PR TITLE
Fixing matching issues in URL plugin

### LINKED ISSUE #4184: support open ip url
**Is your feature request related to a problem? Please describe.**
no

**Describe the solution you'd like**

Supporting IP URLs as first-class clickable targets (just like normal domain URLs) seems useful for development and internal network use cases. It may be worth requesting this as an official enhancement so the core handlers recognize and open IP web addresses automatically.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by VictoriousRaptor**
Fixes #4184. A rework of #4185 as it's very tricky to solve all IPv6 related problems so I just leave it.

What's done:
- Private addresses are no longer ignored.
- Test and matching logic refactored.
- IPv6 support ~~(though some edge cases in comment of #4185 are not handled)~~.

### CODE REVIEW

File: Plugins/Flow.Launcher.Plugin.Url/Main.cs [hunk 1]
```diff
@@ -15,19 +16,28 @@ public class Main : IPlugin, IPluginI18n, ISettingProvider
...
-            "(?:\\.(?:[1-9]\\d?|1\\d\\d|2[0-4]\\d|25[0-4]))" +
+            // IPv6 address with optional brackets (brackets required if followed by port)
+            // IPv6 with brackets
+            "(?:\\[(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}\\]|" + // standard IPv6
+            "\\[(?:[0-9a-fA-F]{1,4}:){1,7}:\\]|" + // IPv6 with trailing ::
+            "\\[(?:[0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}\\]|" + // IPv6 compressed
+            "\\[::(?:[0-9a-fA-F]{1,4}:){0,6}[0-9a-fA-F]{1,4}\\]|" + // IPv6 with leading ::
+            "\\[(?:(?:[0-9a-fA-F]{1,4}:){1,6}|:):(?:[0-9a-fA-F]{1,4}:){0,5}[0-9a-fA-F]{1,4}\\]|" + // IPv6 with :: in the middle
+            "\\[::1\\])" + // IPv6 loopback
+            "|" +
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Url/Main.cs]: @VictoriousRaptor instead of a massive regex string, would `Uri.TryCreate` work for our scenarios? (Also `IPAddress.TryParse` for IPs)

Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Url/Main.cs]: > @VictoriousRaptor instead of a massive regex string, would `Uri.TryCreate` work for our scenarios? (Also `IPAddress.TryParse` for IPs)

I did try but they made it more complicated. `Uri.TryCreate` can't deal with strings without a `xxx://` scheme and `IPAddress.TryParse` can't parse strings with a scheme. I didnt figure out how to mix them, but we need both of them parsed.

**Comment by Jack251970** [comment]
@VictoriousRaptor Is it good to go? I know almost nothing about IP, so it is hard for me to review it by myself

**Comment by VictoriousRaptor** [comment]
> @VictoriousRaptor Is it good to go? I know almost nothing about IP, so it is hard for me to review it by myself

I think it's good to go but let's see if @jjw24 has any comment.

File: Plugins/Flow.Launcher.Plugin.Url/Main.cs [hunk 2]
```diff
@@ -60,7 +32,8 @@ public List<Result> Query(Query query)
                         Score = 8,
                         Action = _ =>
                         {
-                            if (!raw.StartsWith("http://", StringComparison.OrdinalIgnoreCase) && !raw.StartsWith("https://", StringComparison.OrdinalIgnoreCase))
+                            // not a recognized scheme, add preferred http scheme
+                            if (!UrlSchemes.Any(scheme => raw.StartsWith(scheme, StringComparison.OrdinalIgnoreCase)))
                             {
                                 raw = GetHttpPreference() + "://" + raw;
                             }
```
[the remark this answers is not in the document]
Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.Url/Main.cs]: @VictoriousRaptor adding IPv6 support we will need to support this URL creation as well so this suggestion is needed to be addressed.

File: Plugins/Flow.Launcher.Plugin.Url/Main.cs [hunk 3]
```diff
@@ -104,19 +74,50 @@ private static string GetHttpPreference()
...
-            if (UrlRegex.Match(raw).Value == raw) return true;
+            var input = raw.Trim();
 
-            if (raw == "localhost" || raw.StartsWith("localhost:") ||
-                raw == "http://localhost" || raw.StartsWith("http://localhost:") ||
-                raw == "https://localhost" || raw.StartsWith("https://localhost:")
-                )
-            {
+            // Exclude numbers (e.g. 1.2345)
+            if (decimal.TryParse(input, System.Globalization.NumberStyles.Any, System.Globalization.CultureInfo.InvariantCulture, out _))
```
[the remark this answers is not in the document]
Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.Url/Main.cs:94]: @VictoriousRaptor this is not worth considering?

[the remark this answers is not in the document]
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Url/Main.cs:94]: > @VictoriousRaptor this is not worth considering?

No. It's funny that AI told me to use invariant culture in previous comments. Use invariant culture can stop parsing when user is typing a number with dot. Users in culture that uses comma as decimal separator won't have this issue.

File: Plugins/Flow.Launcher.Plugin.Url/Main.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Url/Main.cs]: > @VictoriousRaptor adding IPv6 support we will need to support this URL creation as well so this suggestion is needed to be addressed.

Good catch and will fix later.

File: Flow.Launcher.Test/Plugins/UrlPluginTest.cs [hunk 4]
```diff
@@ -1,33 +1,73 @@
...
+        [TestCase("fe80:1:2::3:4")]
+        [TestCase("::1")]
+        [TestCase("HTTP://EXAMPLE.COM")]
+        [TestCase("HTTPS://EXAMPLE.COM")]
+        [TestCase("EXAMPLE.COM")]
+        [TestCase("LOCALHOST")]
+        public void WhenValidUrlThenIsUrlReturnsTrue(string url)
+        {
+            Assert.That(plugin.IsURL(url), Is.True);
+        }
```
[the remark this answers is not in the document]
Comment (VictoriousRaptor) [reply Flow.Launcher.Test/Plugins/UrlPluginTest.cs:72]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/Flow-Launcher/Flow.Launcher/pull/4191#discussion_r2854269380)

File: Plugins/Flow.Launcher.Plugin.Url/Main.cs [hunk 5]
```diff
@@ -1,56 +1,37 @@
...
+                return [];
+            }
+
+            if (IPEndPoint.TryParse(raw, out var endpoint))
+            {
+                if (endpoint.AddressFamily == System.Net.Sockets.AddressFamily.InterNetworkV6)
+                {
+                    // Enclose IPv6 addresses in brackets for URL formatting
+                    raw = $"[{raw}]";
+                }
```
[the remark this answers is not in the document]
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Url/Main.cs:31]: Resolved in 9017ce6b9d150fcb4385fa3bf00e2cf8ca6ebb3b

File: Plugins/Flow.Launcher.Plugin.Url/Main.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Url/Main.cs]: > @VictoriousRaptor adding IPv6 support we will need to support this URL creation as well so this suggestion is needed to be addressed.

@jjw24 Fixed and added this case to unit test. Please review.

File: Plugins/Flow.Launcher.Plugin.Url/Main.cs [hunk 6]
```diff
@@ -1,56 +1,37 @@
...
-            // port number
-            "(?::\\d{2,5})?" +
-            // resource path
-            "(?:/\\S*)?" +
-            "$";
-        private readonly Regex UrlRegex = new(UrlPattern, RegexOptions.Compiled | RegexOptions.IgnoreCase);
         internal static PluginInitContext Context { get; private set; }
         internal static Settings Settings { get; private set; }
 
+        private static readonly string[] UrlSchemes = ["http://", "https://", "ftp://"];
```
Comment (DavidGBrett) [inline Plugins/Flow.Launcher.Plugin.Url/Main.cs]: As far as I can tell browsers no longer support the ftp protocol for security reasons

- https://developer.chrome.com/blog/chrome-88-deps-rems#ftp_support_removed
- https://blog.mozilla.org/security/2021/07/20/stopping-ftp-support-in-firefox-90/

Probably worth removing support for that in this case, though this may be beyond the scope of this PR

Beyond that it might be worth supporting more scheme types that work for browsers

Of note is internal addresses like ``chrome://`` which is requested in issue #4014 
The equivalent for firefox is ``about:`` I believe
``chrome-extension://`` is another of that kind

Beyond those internal addresses we also have ones like ``view-source:`` and ``file://``

Though supporting these could also be considered out of scope and done in a subsequent PR

**Review by DavidGBrett — COMMENTED** [review]
Definitely looks cleaner than the previous regex solution

I mention some concerns about the UrlSchemes list but that can be set aside for a follow up PR as thats the existing behavior, if prefered

Beyond that it's just a few more trivial tests that could be added, paticularly for fragments

File: Flow.Launcher.Test/Plugins/UrlPluginTest.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (DavidGBrett) [reply Flow.Launcher.Test/Plugins/UrlPluginTest.cs:72]: Query params are tested here but adding some fragments would be good @VictoriousRaptor

File: Plugins/Flow.Launcher.Plugin.Url/Main.cs [hunk 7]
```diff
@@ -104,19 +83,68 @@ private static string GetHttpPreference()
...
+            var urlToValidate = UrlSchemes.Any(s => input.StartsWith(s, StringComparison.OrdinalIgnoreCase))
+                ? input
+                : GetHttpPreference() + "://" + input;
+
+            if (!Uri.TryCreate(urlToValidate, UriKind.Absolute, out var uri))
+                return false;
+            
+
+            // Validate protocol
+            if (uri.Scheme != Uri.UriSchemeHttp && uri.Scheme != Uri.UriSchemeHttps && uri.Scheme != Uri.UriSchemeFtp)
```
Comment (DavidGBrett) [inline Plugins/Flow.Launcher.Plugin.Url/Main.cs]: Ideally this should be based on UrlSchemes instead of repeating it here

File: Flow.Launcher.Test/Plugins/UrlPluginTest.cs [hunk 8]
```diff
@@ -1,33 +1,73 @@
...
+        [TestCase("://example.com")]
+        [TestCase("0.0.0.0")] // Pattern excludes 0.0.0.0
+        [TestCase("256.1.1.1")] // Invalid IPv4
+        [TestCase("example")] // No TLD
+        [TestCase(".com")]
+        [TestCase("http://.com")]
+        public void WhenInvalidUrlThenIsUrlReturnsFalse(string url)
+        {
+            Assert.That(plugin.IsURL(url), Is.False);
         }
```
[the remark this answers is not in the document]
Comment (DavidGBrett) [reply Flow.Launcher.Test/Plugins/UrlPluginTest.cs:93]: Should be addressed in a9be19b9da1b35fbb4b54e7376285f7475701f98

File: Flow.Launcher.Test/Plugins/UrlPluginTest.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (DavidGBrett) [reply Flow.Launcher.Test/Plugins/UrlPluginTest.cs:72]: Should be addressed in a9be19b9da1b35fbb4b54e7376285f7475701f98

File: Flow.Launcher.Test/Plugins/UrlPluginTest.cs [hunk 9]
```diff
@@ -1,33 +1,73 @@
...
+        [TestCase("https://[2001:db8::1]:8080")]
+        [TestCase("[::1]")]
+        [TestCase("[::1]:8080")]
+        [TestCase("2001:db8::1")]
+        [TestCase("fe80:1:2::3:4")]
+        [TestCase("::1")]
+        [TestCase("HTTP://EXAMPLE.COM")]
+        [TestCase("HTTPS://EXAMPLE.COM")]
+        [TestCase("EXAMPLE.COM")]
+        [TestCase("LOCALHOST")]
```
[the remark this answers is not in the document]
Comment (DavidGBrett) [reply Flow.Launcher.Test/Plugins/UrlPluginTest.cs:51]: Should be addressed in a9be19b9da1b35fbb4b54e7376285f7475701f98

File: Plugins/Flow.Launcher.Plugin.Url/Main.cs (hunk 7 above)
Comment (DavidGBrett) [reply Plugins/Flow.Launcher.Plugin.Url/Main.cs]: resolved in fe23bfe

**Review by DavidGBrett — APPROVED** [review]
Leaving aside my concerns about the supported schemes,

I went ahead and commited the minor improvements I saw as required, 
So if you have no problem with what I added on your end then I'm happy to approve!
To be sure I have disabled auto-merge.

Thank you for your contribution!
